test/unittest/memory/unittest_memory_pool.cpp

```diff
@@ -0,0 +1,544 @@
+  EXPECT_EQ(pool.minMemoryRequirement(), 2);
+
+  /** ending overlap with matching ends */
+  pool.requestMemory(3, 10, 12);
+  EXPECT_EQ(pool.minMemoryRequirement(), 3);
+
+  pool.planLayout(nntrainer::BasicPlanner());
+  EXPECT_EQ(pool.minMemoryRequirement(), 3);
+
+  /** ending overlap with matching ends */
+  pool.requestMemory(1, 12, 13);
+  EXPECT_EQ(pool.minMemoryRequirement(), 3);
+
+  pool.planLayout(nntrainer::BasicPlanner());
+  EXPECT_EQ(pool.minMemoryRequirement(), 3);
+}
+
+/**
+ * @brief get memory
+ */
+TEST(MemoryPool, get_memory_01_n) {
+  nntrainer::MemoryPool pool;
+
+  EXPECT_THROW(pool.getMemory(1), std::invalid_argument);
+}
+
+/**
+ * @brief get memory
+ */
+TEST(MemoryPool, get_memory_02_n) {
+  nntrainer::MemoryPool pool;
+
+  auto idx = pool.requestMemory(1, 4, 5);
+  EXPECT_THROW(pool.getMemory(idx), std::invalid_argument);
+}
+
+/**
+ * @brief get memory
+ */
+TEST(MemoryPool, get_memory_03_n) {
+  nntrainer::MemoryPool pool;
+
+  auto idx = pool.requestMemory(1, 4, 5);
+  EXPECT_NO_THROW(pool.planLayout(nntrainer::BasicPlanner()));
+  EXPECT_NO_THROW(pool.allocate());
+
+  EXPECT_ANY_THROW(pool.getMemory(idx + 1));
+
+  EXPECT_NO_THROW(pool.deallocate());
+}
+
+/**
+ * @brief get memory
+ */
+TEST(MemoryPool, get_memory_04_p) {
+  nntrainer::MemoryPool pool;
+  void *mem;
+
+  auto idx = pool.requestMemory(1, 4, 5);
+  EXPECT_NO_THROW(pool.planLayout(nntrainer::BasicPlanner()));
+  EXPECT_NO_THROW(pool.allocate());
+
+  EXPECT_NO_THROW(mem = pool.getMemory(idx));
+  EXPECT_NE(mem, nullptr);
+
+  EXPECT_NO_THROW(pool.deallocate());
+}
+
+/**
+ * @brief validate memory full overlap
+ */
+TEST_P(MemoryPlannerValidate, validate_memory_full_overlap) {
+  nntrainer::MemoryPool pool;
+  std::mt19937 rng;
+  std::uniform_int_distribution<size_t> dist(1, MEM_BYTES);
+  std::uniform_int_distribution<unsigned int> dist_interval(1, INTERVAL_SIZE);
+
+  std::vector<unsigned int> tokens(MEM_QUANT);
+  std::vector<size_t> memory_size(MEM_QUANT);
+  std::vector<void *> ptrs(MEM_QUANT);
+
+  for (unsigned int idx = 0; idx < MEM_QUANT; idx++) {
+    memory_size[idx] = dist(rng);
+    unsigned int start = 1;
+    unsigned int end = start + dist_interval(rng);
+    EXPECT_NO_THROW(tokens[idx] =
+                      pool.requestMemory(memory_size[idx], start, end));
+  }
+
+  EXPECT_NO_THROW(pool.planLayout(*planner.get()));
+  EXPECT_EQ(pool.size(),
+            std::accumulate(memory_size.begin(), memory_size.end(), 0));
+  EXPECT_NO_THROW(pool.allocate());
+
+  for (unsigned int idx = 0; idx < MEM_QUANT; idx++)
+    EXPECT_NO_THROW(ptrs[idx] = pool.getMemory(tokens[idx]));
+
+  /** write data to memory */
+  for (unsigned int idx = 0; idx < MEM_QUANT; idx++)
+    memset(ptrs[idx], idx, memory_size[idx]);
+
+  /** verify data in memory */
+  for (unsigned int idx = 0; idx < MEM_QUANT; idx++) {
+    std::vector<unsigned char> golden(memory_size[idx], idx);
+    memcmp(ptrs[idx], &golden[0], memory_size[idx]);
```

Comment:
does this test holds if it's not basic memory planner?

Reply:
For now, this only holds for basic planner (where the purpose is to check that the memory overlaps). 
This test is updated in the later PR.